### PR TITLE
Zephyr multicore batch 1

### PR DESCRIPTION
**Pull request by lyakh**
The first (IIRC) batch of SOF Zephyr multicore patches. Limited functionality: on APL and CNL booting, secondary core can be powered on and initialised, topology can be loaded, but DMIC isn't working even on core 0, actual audio streaming on core 1, e.g. to HDMI isn't working yet either.

### CODE REVIEW

File: src/platform/Kconfig [hunk 1]
```diff
@@ -227,6 +226,7 @@ config MAX_CORE_COUNT
 
 config CORE_COUNT
 	int "Number of cores"
+	default MP_NUM_CPUS if KERNEL_BIN_NAME = "zephyr"
```
Comment (lgirdwood) [inline src/platform/Kconfig:229]: Btw, are we aligned with all the Zephyr Kconfig symbols ?

File: src/schedule/timer_domain.c [hunk 2]
```diff
@@ -119,6 +119,7 @@ static int timer_domain_register(struct ll_schedule_domain *domain,
 #ifdef __ZEPHYR__
 	void *stack;
 	char *qname;
+	struct k_thread *thread;
```
Comment (lgirdwood) [inline src/schedule/timer_domain.c:122]: I'm assuming that 
1) all threads are idle when there is no work i.e. we are in WFI
2) EDF work is preemptable by LL pipeline tasks ?

I also want to confirm that

3) All IPC is premptable (including via IDC)
4) Pipeline work should mostly be non pre-emptable, but we should also support pre-emtable pipelines and components. i.e. KWD media decode.
5) trace work will be pre-emptable.

**Review by lgirdwood — APPROVED** [review]
LGTM, btw how much stack does each thread take here for LL and EDF ?

**Comment by lgirdwood** [comment]
@lyakh looks like we also have a linking error - see CI
```
238/246] Linking C executable zephyr/zephyr_prebuilt.elf
FAILED: zephyr/zephyr_prebuilt.elf zephyr/zephyr_prebuilt.map 
: && ccache /opt/toolchains/zephyr-sdk-0.12.4/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/xtensa-zephyr-elf-gcc   zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj -o zephyr/zephyr_prebuilt.elf  -Wl,-T  zephyr/linker_zephyr_prebuilt.cmd  -Wl,-Map=/workdir/zephyrproject/build-apl/zephyr/zephyr_prebuilt.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/arch/arch/xtensa/core/startup/libarch__xtensa__core__startup.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/soc/xtensa/intel_adsp/common/libintel_adsp_common.a  modules/sof/libmodules_sof.a  modules/xtensa/libmodules_xtensa_hal.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -L"/opt/toolchains/zephyr-sdk-0.12.4/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/10.2.0"  -L/workdir/zephyrproject/build-apl/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  zephyr/soc/xtensa/intel_adsp/common/libintel_adsp_common.a  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -no-pie  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /workdir/zephyrproject/build-apl/zephyr && /usr/local/bin/cmake -E echo
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build /workdir/zephyrproject/build-apl
/opt/toolchains/zephyr-sdk-0.12.4/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/10.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(timer_domain.c.obj):(.literal.timer_domain_register+0x14): undefined reference to `k_thread_cpu_mask_clear'
/opt/toolchains/zephyr-sdk-0.12.4/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/10.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(timer_domain.c.obj):(.literal.timer_domain_register+0x18): undefined reference to `k_thread_cpu_mask_enable'
/opt/toolchains/zephyr-sdk-0.12.4/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/10.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(timer_domain.c.obj): in function `k_thread_suspend':
/workdir/zephyrproject/build-apl/zephyr/include/generated/syscalls/kernel.h:263: undefined reference to `k_thread_cpu_mask_clear'
/opt/toolchains/zephyr-sdk-0.12.4/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/10.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(timer_domain.c.obj): in function `timer_domain_register':
/workdir/src/schedule/timer_domain.c:170: undefined reference to `k_thread_cpu_mask_enable'
/opt/toolchains/zephyr-sdk-0.12.4/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/10.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(edf_schedule.c.obj): in function `k_thread_suspend':
/workdir/zephyrproject/build-apl/zephyr/include/generated/syscalls/kernel.h:263: undefined reference to `k_thread_cpu_mask_clear'
/opt/toolchains/zephyr-sdk-0.12.4/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/10.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(edf_schedule.c.obj): in function `scheduler_init_edf':
/workdir/zephyr/edf_schedule.c:107: undefined reference to `k_thread_cpu_mask_enable'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by lyakh** [comment]
> LGTM, btw how much stack does each thread take here for LL and EDF ?

@lgirdwood for P4WQ threads, used for IDC I allocate `SOF_STACK_SIZE` which is 4K everywhere ATM

File: src/platform/Kconfig (hunk 1 above)
Comment (lyakh) [reply src/platform/Kconfig:229]: @lgirdwood sorry, what exactly do you mean by "aligned?" We still have a few duplicating Kconfig symbols, there has been no progress on that. But at least I'm trying to synchronise all the multicore related symbols - number of cores (used), multicore support on or off, with these patches you shouldn't be getting any meaningless (or at least confusing) configurations like SMP=y MULTICORE=n etc.

File: src/schedule/timer_domain.c (hunk 2 above)
Comment (lyakh) [reply src/schedule/timer_domain.c:122]: 1. (idle) - should be, yes
2. I think so. Under Zephyr both are implemented as tasks, but the timer domain (for LL scheduling) uses `-CONFIG_NUM_COOP_PRIORITIES` (e.g. -16) as priority, whereas EDF uses 1. Natively with XTOS the LL timer domain handler is run in the timer interrupt context, so, with Zephyr the highest available priority is used.
3. "real" IPC (host->DSP) is unchanged, seems to be mapped to Zephyr task priority 0. IDC runs in P4WQ threads and those use a fixed priority, which is the same as for LL scheduling above - the highest priority of `-CONFIG_NUM_COOP_PRIORITIES`. I guess we have to discuss this...
4. Timer domain pipelines are threads with the highest priority ATM, I think the DMA domain LL tasks are so far unchanged, i.e. called in the DMA interrupt context, but they probably have never been tested with Zephyr, so might be broken.
5. Not sure about Zephyr printk / logging. I think they use some threaded context, e.g. for logging I see a thread created with the `K_LOWEST_APPLICATION_THREAD_PRIO` priority, which is... rather low (depends on Kconfig). But `printk()` seems to run synchronously.

In general, this patch set isn't touching much of the above, it only does one thing WRT threading / priorities: implementing IDC via P4WQ, and there the present API doesn't allow us to select priority. If we need that, we have to discuss that with Zephyr.

**Comment by lyakh** [comment]
CI Zephyr build failure is due to a dependency on my commit https://github.com/zephyrproject-rtos/zephyr/pull/33963/commits/71b975cf8b941ebf17e1c0e01747468db8cca27b

**Comment by lgirdwood** [comment]
> CI Zephyr build failure is due to a dependency on my commit [zephyrproject-rtos/zephyr@71b975c](https://github.com/zephyrproject-rtos/zephyr/commit/71b975cf8b941ebf17e1c0e01747468db8cca27b)

Ok, so we need to get this committed in Zephyr first.

**Comment by lyakh** [comment]
> first

@lgirdwood merged now

**Comment by lyakh** [comment]
SOFCI TEST

**Comment by lyakh** [comment]
The CI build failure: looks like the CI is testing a specific commit ID, not the tip of the zephyr "master." Has to be upgraded

**Review by jgunn0262 — APPROVED** [review]
@lyakh thanks for doing this work, the P4WQ looks very interesting for audio workloads.

**Comment by marc-hb** [comment]
> The CI build failure: 

Always copy/paste the URL of the failure because they tend to become hard to find after force-pushes (which github supports very poorly)
https://github.com/thesofproject/sof/pull/4123/checks?check_run_id=2492951627

> looks like the CI is testing a specific commit ID, not the tip of the zephyr "master." Has to be upgraded

I don't see how it could test a specific commit or which one, does the failure make you suspect that and why?

For sure the build should print which commit it's building, working on that now.

Just for the record this other, more recent and unrelated PR built with Zephyr just fine: https://github.com/thesofproject/sof/pull/4137/checks?check_run_id=2504351596

**Comment by marc-hb** [comment]
https://github.com/thesofproject/sof/pull/4138/checks?check_run_id=2505335361 shows that the script downloads the latest Zephyr version.

**Comment by lyakh** [comment]
> I don't see how it could test a specific commit or which one, does the failure make you suspect that and why?

@marc-hb the reason why I thought it was using an out-of-date Zephyr tree is that these errors:
```
[141/246] Building C object modules/sof/CMakeFiles/modules_sof.dir/workdir/src/schedule/timer_domain.c.obj
/workdir/src/schedule/timer_domain.c: In function 'timer_domain_register':
/workdir/src/schedule/timer_domain.c:170:2: warning: implicit declaration of function 'k_thread_cpu_mask_clear' [-Wimplicit-function-declaration]
  170 |  k_thread_cpu_mask_clear(thread);
      |  ^~~~~~~~~~~~~~~~~~~~~~~
```
and
```
/opt/toolchains/zephyr-sdk-0.12.4/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/10.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(timer_domain.c.obj):(.literal.timer_domain_register+0x14): undefined reference to `k_thread_cpu_mask_clear'
```
in https://github.com/thesofproject/sof/pull/4123/checks?check_run_id=2492951627#step:4:489 and https://github.com/thesofproject/sof/pull/4123/checks?check_run_id=2492951627#step:4:603
respectively would be perfectly clear without the recently merged https://github.com/zephyrproject-rtos/zephyr/pull/33963/commits/c25ff2ce4cbc11b8534ccbb78b71b6be35bc52d6 but aren't clear otherwise. I'm trying to use the script to build this PR locally, if that works, I'll try re-running CI here.

**Comment by lyakh** [comment]
local builds worked. re-running...

**Comment by lyakh** [comment]
SOFCI TEST

**Comment by lyakh** [comment]
@marc-hb can it be, that the zephyr test isn't re-run when a repeated test is forced? I see the same test ID as before my last re-run... That would explain the failure too - the dependency Zephyr PR wasn't yet merged when this PR was first submitted.

**Comment by marc-hb** [comment]
Only Jenkins cares about SOFCI TEST. For all the others you need a SHA change to trigger a new build. `git commit --amend` to change the (date and) SHA1 and then force-push.

**Comment by marc-hb** [comment]
> I see the same test ID as before my last re-run..

You should also check the dates, search for `version.cmake` or `UTC` in the logs.

**Comment by lyakh** [comment]
ok, only checkpatch failing now on some false-positives like the use of c-99 comments for licence strings

**Pull request merged**